File: problem4/vers_py/p4.py
### p4.py
#
# Zeke Reyna
#
# Finds the largest palindrome that is a product
# of two nums between two bounds
import sys

def check_if_pal( i, j ):

    prod = str( i * j )
    half_length = int( len( prod ) / 2 )

    if len( prod ) % 2 == 0:
        front = prod[ : half_length ]
        end   = prod[ half_length : ]
    else:
        front = prod[ : half_length ]
        end   = prod[ half_length +  1 : ]

    end = end[ : : -1 ]

    if front == end:
        retval = i * j
    else:
        retval = -1

    return retval


def main( argv ):

    if len(argv) == 3:
        upperbound = int( argv[2] )
        lowerbound = int( argv[1] )
    else:
        upperbound = 1000
        lowerbound = 100

    largest_pal = 0

    for i in range( lowerbound, upperbound ):
        for j in range( i, upperbound ):

            curr_pal = check_if_pal( i, j )
            if( curr_pal > largest_pal ):
                largest_pal = curr_pal

    print( "\nThe largest parlindrome ",
          "that is a sum of two products ",
          "between the bounds ", lowerbound,
          " and ", upperbound, " is ",
          largest_pal, "\n" )


main( sys.argv )

File: problem4/p4.c
/*
 * p4.c
 *
 * Zeke Reyna
 *
 * A simple C file to find the largest palindrome
 * That's a product of two 3 digit numbers.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SIZE 100 // assuming the string rep of the # is smaller than 100
#define LOWERBOUND 100
#define UPPERBOUND 1000

int check_if_pal(int i, int j)
{
	int z;
	int retval = -1;
	int product = i * j;
	char str[SIZE], front[SIZE], end[SIZE];
	int half_length;

	sprintf(str, "%d", product);
	half_length = strlen(str) / 2;

	if (strlen(str) % 2 == 0) {
		strncpy(front, str, half_length);
		strncpy(end, str + half_length, half_length);

	} else {
		strncpy(front, str, half_length);
		strncpy(end, str + half_length + 1, half_length);
	}
	end[half_length] = '\0';

	retval = i * j;
	for (z = 0; z < half_length; z++) {
		if (front[z] != end[half_length - 1 - z])
			retval = -1;
	}

	return retval;
}

int main( int argc, char **argv )
{
	int i, j;
	int cur_pal;
	int largest_pal = 0;
	int lowerbound, upperbound;

	if (argc == 3) {
		upperbound = atoi(argv[2]);
		lowerbound = atoi(argv[1]);
	} else {
		upperbound = UPPERBOUND;
		lowerbound = LOWERBOUND;
	}

	for (i = lowerbound; i < upperbound; i++) {
		for (j = i; j < upperbound; j++) {

			cur_pal = check_if_pal(i, j);
			if (cur_pal > largest_pal)
				largest_pal = cur_pal;
		}
	}

	printf( "\nThe largest palindrome " );
	printf( "that is a sum of two products between " );
	printf( "the bounds %d and %d is: ", lowerbound, upperbound );
	printf( "%d\n\n", largest_pal );

	return 0;
}
